src/main/kotlin/com/fasterxml/jackson/module/kotlin/KotlinAnnotationIntrospector.kt

```diff
@@ -59,6 +61,42 @@ internal class KotlinAnnotationIntrospector(private val context: Module.SetupCon
         return super.findCreatorAnnotation(config, a)
     }
 
+    // Find a serializer to handle the case where the getter returns an unboxed value from the value class.
+    override fun findSerializer(am: Annotated): ValueClassBoxSerializer? = when (am) {
+        is AnnotatedMethod -> {
+            val getter = am.member.apply {
+                // If the return value of the getter is a value class,
+                // it will be serialized properly without doing anything.
+                if (this.returnType.isUnboxableValueClass()) return null
+            }
+
+            val kotlinProperty = getter
+                .declaringClass
+                .kotlin
+                .let {
+                    // KotlinReflectionInternalError is raised in GitHub167 test,
+                    // but it looks like an edge case, so it is ignored.
+                    try {
+                        it.memberProperties
+                    } catch (e: Error) {
+                        null
+                    }
+                }?.find { it.javaGetter == getter }
+
+            (kotlinProperty?.returnType?.classifier as? KClass<*>)
+                ?.takeIf { it.isValue }
+                ?.let { outerClazz ->
+                    @Suppress("UNCHECKED_CAST")
+                    ValueClassBoxSerializer(outerClazz as KClass<Any>, getter.returnType)
+                }
+        }
+        // Ignore the case of AnnotatedField, because JvmField cannot be set in the field of value class.
+        else -> null
+    }
+
+    // Perform proper serialization even if the value wrapped by the value class is null.
+    override fun findNullSerializer(am: Annotated) = findSerializer(am)
```

Comment:
Does this actually work as expected? Call should only be made if the value class itself is `null`; not for wrapped value. I may be wrong since I am not familiar with types here, just suggest validation of all cases.

Reply:
If the property itself is `null`, the return value from the `java getter` will be in an `unboxed` state, so the `ValueClassBoxSerializer` will not be used.
This is verified by the following comparison.
 https://github.com/FasterXML/jackson-module-kotlin/pull/527/files#diff-7c6b0a8c396b759fd146df36577c7a8fb3cbe94406e6912a729f6e028413411bR49-R51

Reply:
Ok. As long as unit tests cover expected behavior it's probably fine.

One other note: "null serializer" is only ever used for writing `null` value or placeholder, and never anything more advanced. So one typically returns sort of special constant value serializing implementation, instead of delegating.

Reply:
@cowtowncoder @tatu-at-datastax 
Sorry if I'm not reading this correctly.

The `findNullSerializer` returns a serializer for values that are `non-null` in `Kotlin`.
I meant to add a comment to explain that.

I think it might have been better to have a `private` function named something like `findSerializerForValueClass` and call the same function in `findSerializer`/`findNullSerializer`.
If you have any comments or implementations that are easier to understand, please let me know.

Reply:
My point is simply this: method `findNullSerializer` is called by Jackson only EVER when there is a literal JVM-level `null` to serialize. Regular serializers are NEVER given `null` values to serialize; they are not expected to (have to) do null checks, nor know what to serialize.

In that sense it seems/seemed odd that implementation would delegate to `findSerializer()`: I am not sure why this is done. It seems rather like it should not be defined at all.

But then again Kotlin module sometimes overrides things I do not expect it to, partially since null handling in Kotlin is quite different from plain Java.

I'll let you and Kotlin module owners decide here what to do; I just try to give input from Jackson databind side to help develop things :)